### PR TITLE
Add HtmlElementView (the Flutter Web platform view)

### LINKED ISSUE #37791: [web]Upstream RenderHtmlView
Upstream RenderHtmlView to platform_view.dart to support web platform views.

### LINKED ISSUE #37794: [web][upstream] Upstream HtmlView

### LINKED ISSUE #37795: [web][upstream] Upstream HtmlViewController/initHtmlView

### PR DESCRIPTION
**Pull request by harryterkelsen**
## Description

This adds the `HtmlElementView` to Flutter. This widget is the equivalent of `AndroidView` or `UIKitView` for Flutter Web. 

## Related Issues

* Closes https://github.com/flutter/flutter/issues/37795
* Closes https://github.com/flutter/flutter/issues/37791
* Closes https://github.com/flutter/flutter/issues/37794

## Tests

I added the following tests:

I added tests alongside tests for `AndroidView` and `UIKitView` in `services/platform_views_test.dart` and `widgets/platform_view_test.dart`.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by amirh** [comment]
Hi, it would help me if you could point me to the companion code that sizes / creates / presents webview on the web so I have more context when reviewing.

**Comment by hterkelsen** [comment]
The corresponding platform code is here: https://github.com/flutter/engine/blob/master/lib/stub_ui/lib/src/engine/platform_views.dart

Only `create` is implemented on the platform side so far.

**Comment by amirh** [comment]
Is there a design doc available? Seems like we're going with an approach similar to `AndroidView` for sizing which is more complex than the `UiKitView` approach due to specific Android limitations.

We're working on migrating as much shared functionality from `AndroidView` and `UiKitView` to common classes (`PlatformViewSurface` and `PlatformViewLink`), see https://github.com/flutter/flutter/issues/36779 some of it landed already, I wonder if we can use that for the web as well.

**Comment by hterkelsen** [comment]
I don't have a design doc. Most of the code for `HtmlView` is copied from `AndroidView` but now that I'm looking over it again and I see that some of the things in the view controller remained that I meant to take out (ie `acceptGesture` and `rejectGesture`).

I'll write a doc for why I wasn't able to do gesture detection in the framework for `HtmlView` and also simplify the resizing code and update this PR tomorrow.

**Comment by hterkelsen** [comment]
Updated the code. I'm writing a doc now explaining why web restricts `HtmlView`s from participating in Flutter gesture detection and explaining the `HtmlView` widget lifecycle. I'll update this PR when the doc is ready

**Comment by hterkelsen** [comment]
@amirh I wrote a doc explaining `HtmlView` and why I was unable to do gesture detection for them. Take a look here: https://docs.google.com/document/d/1JxXrxzO2CJGcrDYvTGQqbJDW4S2chLp3vCrzPaZsGHI/edit?usp=sharing

File: packages/flutter/lib/src/widgets/platform_view.dart
Comment (amirh) [inline packages/flutter/lib/src/widgets/platform_view.dart]: I think we should be able to save some code duplication by re-using `PlatformViewLink` and `PlatformViewSurface` here ([PlatformViewLink](https://master-api.flutter.dev/flutter/widgets/PlatformViewLink-class.html)). We can create a controller with a no-op pointer event dispatcher.

`PlatformViewLink` is currently missing support for `layoutDirection` but probably a good idea to add it there anyway.

**Comment by rostopira** [comment]
Any ETA for this PR to merge? It's a blocker for me currently

File: packages/flutter/lib/src/widgets/platform_view.dart
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/platform_view.dart]: This sounds like I can add HTML to any app running on any platform (e.g. Android, iOS, etc.). Is that true? If not, can we clearly document where this is available?

File: packages/flutter/lib/src/rendering/platform_view.dart
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/platform_view.dart]: Document where this is available.

**Comment by chertov** [comment]
I'm waiting for release! It will awesome to get <video/>, Media Source API, webrtc.. etc

**Comment by rostopira** [comment]
On android you can set this parameters on `JavaAudioDeviceModule` directly
https://github.com/cloudwebrtc/flutter-webrtc/blob/master/android/src/main/java/com/cloudwebrtc/webrtc/FlutterWebRTCPlugin.java#L113
Modify source and forget about it, I have high doubts that anyone else would ever need this

File: packages/flutter/lib/src/widgets/platform_view.dart
Comment (rostopira) [inline packages/flutter/lib/src/widgets/platform_view.dart]: Says, that PlatformViewFactory must be registered, but how??

Comment (muhleder) [reply packages/flutter/lib/src/widgets/platform_view.dart]: There's an example in the flutter_web fork here https://github.com/flutter/flutter_web/blob/master/examples/html_platform_view/lib/main.dart and it looks like the corresponding code in the main branch is in the engine at https://github.com/flutter/engine/blob/master/lib/web_ui/lib/src/engine/platform_views.dart

Which you would have to register in a plugin? An example of which for ios is in the webview_flutter plugin here https://github.com/flutter/plugins/blob/560694c5b58eed8a46e57b1dd61e701591209340/packages/webview_flutter/android/src/main/java/io/flutter/plugins/webviewflutter/WebViewFlutterPlugin.java

Not sure what this plugin registration would look like on the web platform though?

**Comment by hterkelsen** [comment]
PTAL. I changed the implementation of `HtmlView` to use `PlatformViewLink`

**Review by amirh — COMMENTED** [review]
The code loos good to me, left a few minor nits.

The one thing I'm hesitant about is the name `HtmlView` (left a comment), let's make sure we settle on a name we all like before landing this.

File: packages/flutter/lib/src/widgets/platform_view.dart
Comment (rostopira) [reply packages/flutter/lib/src/widgets/platform_view.dart]: `IFrameView` will be confusing for users that aren't familiar with web development. `FlutterWebHtmlView` is too long and doesn't make it less ambiguous anyway. I think mobile devs won't be confused by `HtmlView` or misusing it like webview. So vote for that, keep it as it is. As an options, I would suggest `WebPlatformView` or `HtmlPlatformView`

File: packages/flutter/lib/src/widgets/platform_view.dart
Comment (rostopira) [reply packages/flutter/lib/src/widgets/platform_view.dart]: There is `kIsWeb` somewhere. And also `identical(0,0.0)`

**Review by amirh — APPROVED** [review]
LGTM

**Comment by rostopira** [comment]
my-awesome-meme

**Comment by CiriousJoker** [comment]
So now what? When can this be used in flutter_web? So far, it's only in the flutter repo, right?

**Comment by ferhatb** [comment]
Will land in flutter_web shortly.

**Comment by rostopira** [comment]
Could it be used in flutter master channel? Can't find a way to get platform view registry

**Comment by cbenhagen** [comment]
I am currently using it like this:
```dart
import 'dart:html';
import 'dart:ui' as ui;

import 'package:flutter/material.dart';

class WebVideoElement extends StatefulWidget {
  const WebVideoElement({Key key, this.src, this.width, this.height, this.startAt})
      : super(key: key);

  final int width;
  final int height;
  final String src;
  final double startAt;

  @override
  _WebVideoElementState createState() => _WebVideoElementState();
}

class _WebVideoElementState extends State<WebVideoElement> {
  @override
  void initState() {
    super.initState();
    ui.platformViewRegistry.registerViewFactory(
        widget.src,
        (int viewId) {
          final video = VideoElement()
            ..width = widget.width
            ..height = widget.height
            ..src = widget.src + '#t=${widget.startAt}'
            ..autoplay = true
            ..controls = true
            ..style.border = 'none';
          return video ;
        });
  }

  @override
  Widget build(BuildContext context) {
    return Directionality(
      textDirection: TextDirection.ltr,
      child: HtmlElementView(viewType: widget.src),
    );
  }
}
```

I was not sure how I could make the dart analyzer recognize dart:html stuff so I excluded the file for now.

**Comment by rostopira** [comment]
@cbenhagen Thank you

**Comment by cbenhagen** [comment]
@rostopira I am using this around the other widget to control the size:

```dart
class WebVideoPlayer extends StatelessWidget {
  const WebVideoPlayer({Key key, this.src, this.startAt}) : super(key: key);

  final String src;
  final double startAt;

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(
      builder: (context, constraints) {
        return WebVideoElement(
          width: constraints.maxWidth.round(),
          height: constraints.maxHeight.round(),
          src: src,
          startAt: startAt,
        );
      },
    );
  }
}
```

@hterkelsen is there a better way to have a `HtmlElementView` dynamically adjust its size?

**Comment by AppleEducate** [comment]
> Could it be used in flutter master channel? Can't find a way to get platform view registry

I am facing this too. I cannot find out how to register an element. I could do it on `flutter_web` fork but cannot find `platformViewRegistry.registerViewFactory`.

 Am i missing something?

**Comment by rostopira** [comment]
@AppleEducate just ignore this error, dart analyzer doesn't know about this property, but it exists in compile time

**Comment by AppleEducate** [comment]
ok perfect, thank you!

**Comment by NicolasDionB** [comment]
Any idea where I could get documentation on this, besides in the code itself? Examples? Thanks!

**Comment by bjayaram** [comment]
I tried to embed the WebVideoElement shared by @cbenhagen in a popup window but the video doesn't load. I see an error in the dev console like so:

```
Uncaught Error: Assertion failed: org-dartlang-app:///packages/flutter/src/gestures/converter.dart:132:20
!state.down
is not true
    at Object.throw_ [as throw] (:62751/packages/build_web_compilers/src/dev_compiler/dart_sdk.js:4010)
    at Object.assertFailed (:62751/packages/build_web_compilers/src/dev_compiler/dart_sdk.js:3956)
    at SyncIterable.new.expand (:62751/packages/flutter/src/gestures/arena.ddc.js:9738)
    at expand.next (<anonymous>)
    at ListQueue.new.addAll (:62751/packages/build_web_compilers/src/dev_compiler/dart_sdk.js:45558)
    at binding$5.WidgetsFlutterBinding.new.[_handlePointerDataPacket] (:62751/packages/flutter/src/gestures/arena.ddc.js:9284)
    at _engine.PointerBinding.new.[_onPointerData] (:62751/packages/build_web_compilers/src/dev_compiler/dart_sdk.js:143014)
    at :62751/packages/build_web_compilers/src/dev_compiler/dart_sdk.js:143215
    at HTMLElement.<anonymous> (:62751/packages/build_web_compilers/src/dev_compiler/dart_sdk.js:143160)
```
Not sure if I need to file a different issue or post here.
Would appreciate any any clue as to how to fix it. 
Thanks.

**Comment by RobinJespersen** [comment]
Hi,
I am using the HtmlElementView to show an iFrame.
(like here: https://medium.com/flutter-community/flutter-web-and-iframe-f26399aa1e2a)
Except that I have placed it in a Scaffold.

But as soon as a Tooltip is shown, the iFrame jumps back to its initial state.
So any scroll change or url change is being lost.

Disabling the Tooltips "solves" the problem.

I have no more Ideas how to debug that.
So if someone noticed the same behavior or has an Idea how to fix that, I would appreciate any feedback.

Thanks

**Comment by RobinJespersen** [comment]
update, my problem is solved on the dev channel...

**Pull request merged**
